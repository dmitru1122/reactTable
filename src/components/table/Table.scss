@import '../../theme/variables.scss';
.table {
  .table {
    &__row {
      cursor: pointer;
      &__cell {
        &--spinner {
          text-align: center !important;
        }
        &--short {
          cursor: initial;
          width: 15rem;
          @media screen and (max-width: $breakpoint-lg) {
            width: 10rem;
          }
        }
      }
    }
  }
}
