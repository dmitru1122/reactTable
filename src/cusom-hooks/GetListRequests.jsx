import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import actionTypes from '../redux/actions/actionTypes';

export function useGetRequestData() {
  const requestsData = useSelector((state) => state.requestData);
  const lastUpdate = useSelector((state) => state.lastUpdate);
  const dispatch = useDispatch();
  const [data, setData] = useState(null);

  useEffect(() => {
    if (requestsData?.headerData) {
      setData(requestsData);
    } else {
      dispatch({ type: actionTypes.LOAD_DATA });
    }
  }, [lastUpdate, requestsData]);

  return data;
}

export const tti = 'sdfsd';
