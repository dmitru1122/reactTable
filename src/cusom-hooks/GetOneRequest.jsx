import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import actionTypes from '../redux/actions/actionTypes';

export function useGetOneRequest(id) {
  const requestsData = useSelector((state) => state.fullRequestInfo);
  const dispatch = useDispatch();
  const [data, setData] = useState(null);

  useEffect(() => {
    if (requestsData[id]) {
      setData(requestsData[id]);
    } else {
      dispatch({ type: actionTypes.LOAD_ONE_REQUEST, id });
    }
  }, [requestsData]);

  return data;
}

export const test = 'dfsf';
