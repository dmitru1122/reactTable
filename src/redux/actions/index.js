import actionTypes from './actionTypes';

export function failure(error) {
  return {
    type: actionTypes.FAILURE,
    error,
  };
}

export function loadData() {
  return { type: actionTypes.LOAD_DATA };
}

export function loadDataSuccess(data, id) {
  return {
    type: actionTypes.LOAD_DATA_SUCCESS,
    data,
    id,
  };
}

export function addOneRequest(data) {
  return {
    type: actionTypes.ADD_ONE_REQUEST,
    data,
  };
}

// export function addOneRequestSuccess(data) {
//   return {
//     type: actionTypes.ADD_ONE_REQUEST_SUCCESS,
//     data,
//   };
// }
// export function addOneRequestFail() {
//   return {
//     type: actionTypes.ADD_ONE_REQUEST_FAIL,
//   };
// }

export function loadOneRequest(id) {
  return { type: actionTypes.LOAD_ONE_REQUEST, id };
}

export function loadOneRequestSuccess(data, id) {
  return {
    type: actionTypes.LOAD_ONE_REQUEST_SUCCESS,
    data,
    id,
  };
}

export function deleteOneRequest(id) {
  return { type: actionTypes.DELETE_ONE_REQUEST, id };
}
export function deleteOneRequestReset() {
  return { type: actionTypes.DELETE_ONE_REQUEST_RESET };
}

export function deleteOneRequestSuccess(id) {
  return { type: actionTypes.DELETE_ONE_REQUEST_SUCCESS, id };
}
export function deleteOneRequestFail() {
  return { type: actionTypes.DELETE_ONE_REQUEST_FAIL };
}

export function editOneRequest(data, id) {
  return { type: actionTypes.EDIT_ONE_REQUEST, data, id };
}
// export function editOneRequestSuccess(data, id) {
//   return { type: actionTypes.EDIT_ONE_REQUEST_SUCCESS, data, id };
// }
// export function editOneRequestFail() {
//   return { type: actionTypes.EDIT_ONE_REQUEST_FAIL };
// }
