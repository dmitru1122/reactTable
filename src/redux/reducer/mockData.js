export const initialState = {
  error: null,
  lastUpdate: 0,
  requestData: { headerData: null, rowData: null },
  deleteRequestStatus: 'waiting',
  fullRequestInfo: {},
  placeholderData: null,
};

export const rowData = [{ firstName: 'Vit', gender: 'male', id: '2', lastName: 'vab', purpose: 'Job proposition' }];
export const rowDataAfterFilter = [{ firstName: 'Vit', id: '2', purpose: 'Job proposition' }];
export const headerData = [
  { key: 'purpose', header: 'Purpose' },
  { key: 'firstName', header: 'First Name' },
];
export const fullHeaderData = [
  { key: 'purpose', header: 'Purpose' },
  { key: 'firstName', header: 'First Name' },
  { key: 'lastName', header: 'Last Name' },
  { key: 'gender', header: 'Purpose' },
];

export const failureMock = {
  data: {
    headerData,
    rowData: rowDataAfterFilter,
  },
  id: '2',
};

export const loadDataMock = {
  data: {
    headerData,
    rowData,
  },
  id: '2',
};
export const loadOneRequestMock = {
  data: {
    headerData: fullHeaderData,
    rowData,
  },
  id: '2',
};
export const loadSecondOneRequestMock = {
  data: {
    headerData: fullHeaderData,
    rowData: [{ firstName: 'Dima', gender: 'male', id: '3', lastName: 'Pashkevich', purpose: 'Job proposition' }],
  },
  id: '3',
};

export const addOneRequestDataMock = rowData;
//  {
//   firstName: 'Inna',
//   gender: 'male',
//   id: '',
//   lastName: 'test',
//   purpose: 'Job proposition',
// };

export const editOneRequestMock = {
  data: rowData,
  id: '2',
};
