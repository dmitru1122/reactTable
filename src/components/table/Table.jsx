import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import './Table.scss';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  DataTable,
  TableToolbar,
  TableToolbarContent,
  TableToolbarSearch,
  TableContainer,
  Table,
  TableHead,
  TableRow,
  TableHeader,
  TableBody,
  TableCell,
} from 'carbon-components-react';
import { deleteOneRequest, deleteOneRequestReset } from '../../redux/actions/index';
import Modal from '../modal/Modal';
import Notice from '../modal/Notice';
import SpinnerCS from '../spinner-cs/Spinner';

const propTypes = {
  headerData: PropTypes.arrayOf(PropTypes.shape({ key: PropTypes.string, header: PropTypes.string })),
  rowData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      firstName: PropTypes.string,
      purpose: PropTypes.string,
    }),
  ),
};
const defaultProps = {
  headerData: [
    { key: 'purpose', header: 'Purpose' },
    { key: 'firstName', header: 'First Name' },
  ],
  rowData: [{ id: '', firstName: '', purpose: '' }],
};

function TableRequests(props) {
  const dispatch = useDispatch();
  const history = useHistory();
  const { rowData, headerData } = props;
  const [deleting, setDeleting] = useState({ status: 'beforeAction', id: 0 });
  const status = useSelector((state) => state.deleteRequestStatus);

  const openFullRequestInfo = (id) => {
    history.push(`/request/${id}`);
  };
  const handleClickDelete = (id) => {
    setDeleting({ status: 'working', id });
    dispatch(deleteOneRequest(id));
  };
  useEffect(() => {
    let reset = 0;

    if (status === 'resolve') {
      setDeleting({ status: 'success', id: 0 });
      reset = setTimeout(() => {
        dispatch(deleteOneRequestReset());
        setDeleting({ status: false, id: 0 });
      }, 5000);
    }
    if (status === 'reject') {
      setDeleting({ status: 'reject', id: 0 });
      reset = setTimeout(() => {
        dispatch(deleteOneRequestReset());
        setDeleting({ status: false, id: 0 });
      }, 5000);
    }
    return () => {
      clearTimeout(reset);
    };
  }, [status]);
  return (
    <div className='table'>
      <DataTable rows={rowData} headers={headerData} isSortable>
        {({ rows, headers, getHeaderProps, getTableProps, onInputChange, getBatchActionProps }) => (
          <TableContainer title='List of requests'>
            <TableToolbar>
              <TableToolbarContent>
                <TableToolbarSearch
                  tabIndex={getBatchActionProps().shouldShowBatchActions ? -1 : 0}
                  onChange={onInputChange}
                />
                {/* <Button
                  tabIndex={getBatchActionProps().shouldShowBatchActions ? -1 : 0}
                  onClick={() => handleClickEdit('addRequest', '0')}
                  renderIcon={Add}
                  kind='primary'>
                  Add new
                </Button> */}
                <Modal
                  type='add'
                  title='Add request'
                  buttonLabel='I have a component'
                  description='hi'
                  action='Delete'
                />
              </TableToolbarContent>
            </TableToolbar>
            <Table {...getTableProps()}>
              <TableHead>
                <TableRow>
                  {headers.map((header) => (
                    <TableHeader {...getHeaderProps({ header })}>{header.header}</TableHeader>
                  ))}
                  <TableHeader className='bx--table-sort-v2__ico'>Edit</TableHeader>
                </TableRow>
              </TableHead>
              <TableBody>
                {rows.map((row) => (
                  <TableRow key={row.id} className='table__row'>
                    {deleting.status === 'working' && deleting.id === row.id ? (
                      <td id='spin' className='table__row__cell--spinner' colSpan={row.cells.length + 1}>
                        <SpinnerCS />
                      </td>
                    ) : (
                      <>
                        {row.cells.map((cell) => (
                          <TableCell key={cell.id} onClick={() => openFullRequestInfo(row.id)}>
                            {cell.value}
                          </TableCell>
                        ))}
                        <TableCell className='table__row__cell--short'>
                          <Modal
                            type='delete'
                            title='Delete'
                            buttonLabel='I have a component'
                            description='Are you sure you want to go on?'
                            action='Delete'
                            continueAction={() => handleClickDelete(row.id)}
                          />

                          <Modal type='edit' title='Edit row' action='Delete' id={row.id} />
                        </TableCell>
                      </>
                    )}
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        )}
      </DataTable>
      <Notice isShowModal={deleting.status === 'success'} title='Warning' description='Request was deleted' />
      <Notice isShowModal={deleting.status === 'reject'} title='Error' description='Something was wrong' />
    </div>
  );
}

TableRequests.propTypes = propTypes;
TableRequests.defaultProps = defaultProps;

export default TableRequests;
