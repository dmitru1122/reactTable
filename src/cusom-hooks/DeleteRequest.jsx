import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { deleteOneRequest, deleteOneRequestReset } from '../redux/actions/index';

function useDeleteRequest() {
  const [id, setId] = useState();
  // const [status, setStatus] = useState('pending');
  const listRequests = useSelector((state) => state.requestData);
  const stat = useSelector((state) => state.deleteRequestStatus);
  const dispatch = useDispatch();
  console.log('rerender');
  listRequests.rowData.forEach((item) => {
    if (+item.id === +id && stat !== 'reject') {
      console.log('check pass, but it"s wrong');
      dispatch(deleteOneRequest(id));
    }
  });

  useEffect(() => {
    console.log(stat);
    // setStatus(stat);
    setTimeout(() => dispatch(deleteOneRequestReset()), 3000);
  }, [stat]);

  return { status: stat, setId };
}

export default useDeleteRequest;
