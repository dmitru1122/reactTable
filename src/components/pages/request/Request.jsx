import { useParams } from 'react-router-dom';
import { Container, Row, Col } from 'reactstrap';
import { useGetOneRequest } from '../../../cusom-hooks/GetOneRequest';
import ReturnToListButton from '../../buttons/LinkButton';
import LoadTable from '../../load-table/LoadTable';

const Request = () => {
  const params = useParams();
  const requestData = useGetOneRequest(params.id);
  return (
    <main className='home'>
      <Container>
        <Row>
          <Col className='d-flex justify-content-end'>
            <ReturnToListButton link='/' />
          </Col>
        </Row>
      </Container>
      <LoadTable data={requestData} />
    </main>
  );
};

export default Request;
