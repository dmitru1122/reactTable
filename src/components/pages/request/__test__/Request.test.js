import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Spinner from '../../../spinner-cs/Spinner';
import Table from '../../../table/Table';
import Request from '../Request';
import ReturnToListButton from '../../../buttons/LinkButton';
import { initialState, loadOneRequestMock } from '../../../../redux/mockData';
import { useGetOneRequest } from '../../../../cusom-hooks/GetOneRequest';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ id: '2' }),
}));

jest.mock('../../../../cusom-hooks/GetOneRequest');

const setUp = (store) =>
  mount(
    <Provider store={store}>
      <Request />
    </Provider>,
  );

const fullRequestInfo = loadOneRequestMock;

describe('Reqest Page', () => {
  const mockStore = configureStore();
  const store = mockStore(initialState);
  let component;

  describe('loading', () => {
    beforeEach(() => {
      component = setUp(store);
    });

    it('should render cases Reqest page', () => {
      expect(component).toMatchSnapshot();
    });

    it('should render request without crashing', () => {
      expect(component).toBeTruthy();
    });

    it('should render ReturnToListButton', () => {
      const element = <ReturnToListButton />;
      expect(component.contains(element)).toBeTruthy();
    });

    it('should render spinner', () => {
      const element = <Spinner />;
      expect(component.contains(element)).toBeTruthy();
    });
  });

  it('should render Table', () => {
    const element = <Table headerData={fullRequestInfo.data.headerData} rowData={fullRequestInfo.data.rowData} />;

    useGetOneRequest.mockReturnValue({
      headerData: fullRequestInfo.data.headerData,
      rowData: fullRequestInfo.data.rowData,
    });
    const wrapper = setUp(store);
    expect(wrapper.contains(element)).toBeTruthy();
  });
});
