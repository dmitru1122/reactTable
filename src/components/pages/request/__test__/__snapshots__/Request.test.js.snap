// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reqest Page loading should render cases Reqest page 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Request>
    <main
      className="home"
    >
      <Container
        tag="div"
      >
        <div
          className="container"
        >
          <Row
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <div
              className="row"
            >
              <Col
                className="d-flex justify-content-end"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="d-flex justify-content-end col"
                >
                  <LinkButton
                    label="Return to the list requests"
                    link="/"
                  >
                    <Button
                      dangerDescription="danger"
                      data-testid="LinkButton"
                      disabled={false}
                      isExpressive={false}
                      kind="ghost"
                      onClick={[Function]}
                      renderIcon={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "render": [Function],
                        }
                      }
                      size="default"
                      tabIndex={0}
                      tooltipAlignment="center"
                      tooltipPosition="top"
                      type="button"
                    >
                      <button
                        aria-describedby={null}
                        aria-pressed={null}
                        className="bx--btn bx--btn--ghost"
                        data-testid="LinkButton"
                        disabled={false}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        Return to the list requests
                        <ForwardRef(Undo32)
                          aria-hidden="true"
                          className="bx--btn__icon"
                        >
                          <Icon
                            aria-hidden="true"
                            className="bx--btn__icon"
                            fill="currentColor"
                            height={32}
                            preserveAspectRatio="xMidYMid meet"
                            viewBox="0 0 32 32"
                            width={32}
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <svg
                              aria-hidden={true}
                              className="bx--btn__icon"
                              fill="currentColor"
                              focusable="false"
                              height={32}
                              preserveAspectRatio="xMidYMid meet"
                              viewBox="0 0 32 32"
                              width={32}
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M20,10H7.8149l3.5874-3.5859L10,5,4,11,10,17l1.4023-1.4146L7.8179,12H20a6,6,0,0,1,0,12H12v2h8a8,8,0,0,0,0-16Z"
                              />
                            </svg>
                          </Icon>
                        </ForwardRef(Undo32)>
                      </button>
                    </Button>
                  </LinkButton>
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
      <LoadTable
        data={null}
      >
        <div
          className="d-flex align-items-center justify-content-center"
        >
          <SpinnerCs>
            <div
              className="spinner-border text-primary"
              role="status"
            >
              <span
                className="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </SpinnerCs>
        </div>
      </LoadTable>
    </main>
  </Request>
</Provider>
`;
